aws/session/session.go

```diff
@@ -12,87 +12,56 @@ import (
 
 // NewPoolSafe is used to create a pool of AWS sessions with different profile/region permutations
 func NewPoolSafe(profiles []string, regions []string, logger *log.Logger) (allSessions *PoolSafe) {
+	sessions := map[string]*Pool{}
 
-	wg := sync.WaitGroup{}
-	sp := &PoolSafe{
-		Sessions: make(map[string]*Pool),
-	}
-
-	if len(regions) != 0 {
-		for _, p := range profiles {
-			for _, r := range regions {
-				wg.Add(1)
-				// Wait until we have the session for each permutation of profiles and regions
-				go func(p string, r string) {
-					defer wg.Done()
-
-					s, err := newSession(p, r)
-					if err != nil {
-						logger.Fatalf("Error when trying to create session:\n%v", err)
-					}
+	for _, region := range regions {
+		if stsCredentialsSet() {
+			pool := newPool("", region, logger)
+			poolName := fmt.Sprintf("default-%s", region)
+			sessions[poolName] = pool
 
-					if err = validateSessionCreds(s); err != nil {
-						logger.Fatal(err)
-					}
+			continue
+		}
 
-					session := Pool{
-						Logger:      logger,
-						ProfileName: p,
-						Session:     s,
-					}
-					sp.Sessions[fmt.Sprintf("%s-%s", p, r)] = &session
-				}(p, r)
-			}
+		for _, profile := range profiles {
+			pool := newPool(profile, region, logger)
+			poolName := fmt.Sprintf("%s-%s", profile, region)
+			sessions[poolName] = pool
 		}
-	} else {
-		for _, p := range profiles {
-			wg.Add(1)
-			// Wait until we have the session for each profile
-			go func(p string) {
-				defer wg.Done()
+	}
 
-				s, err := newSession(p, "")
-				if err != nil {
-					logger.Fatalf("Error when trying to create session:\n%v", err)
-				}
+	return &PoolSafe{Sessions: sessions}
+}
 
-				if err = validateSessionCreds(s); err != nil {
-					logger.Fatal(err)
-				}
+func stsCredentialsSet() bool {
+	sessionToken := os.Getenv("AWS_SESSION_TOKEN")
+	accessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
+	secretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
 
-				session := Pool{
-					Logger:      logger,
-					ProfileName: p,
-					Session:     s,
-				}
-				sp.Sessions[fmt.Sprintf("%s", p)] = &session
-			}(p)
-		}
+	if sessionToken == "" || accessKeyID == "" || secretAccessKey == "" {
+		return false
 	}
-	// Wait until all sessions have been initialized
-	wg.Wait()
 
-	return sp
+	return true
 }
 
-func validateSessionCreds(session *session.Session) (err error) {
```

Comment:
This stops attempting to validate the credentials in this way. This was hard to test/mock out. For profile configuration we have the list of profiles that exist for a user, we can easily validate the passed profile is in that list. As for validating the credentials are correct for both STS credentials and profile credentials we should just appropriately handle that failure where we use the client and surface a message the clearly communicates the issue

Reply:
I couldn't initially remember how things would behave if invalid credentials are provided, but I realize now it was due to how an invalid profile name + the credential provider chain + finto interacts. In hindsight, we should be fine to remove this because the dialer timeout is now explicitly set to an appropriate value.